### PR TITLE
reconnect only on fail ping

### PR DESCRIPTION
**Pull request by snwfog**
Addressing: https://github.com/go-redis/redis/issues/1518

### CODE REVIEW

**Comment by rbg** [comment]
can you describe what you think your code change is doing please.

**Comment by snwfog** [comment]
@rbg  I am trying to solve few issues mentioned in #1518. 

1. On the first timeout, `pingErr` is not used at all from calling `Ping`.
2. On a consecutive timeout, in the case where `pingErr` is nil, we still close current connection and reconnect. This has lead to lost of messages.
3. `healthy` is used locally only; it doesn't need to be inside of mutex.

**Comment by monkey92t** [comment]
How about this way? Any suggestions?

```go
type PubSub struct {
	// ...
	healthTimeout time.Duration
}

func (c *PubSub) init() {
	//...
	if c.healthTimeout == 0 {
		c.healthTimeout = 5 * time.Second
	}
}

func (c *PubSub) SetHealthTimeout(t time.Duration) {
	if t > 0 {
		c.healthTimeout = t
	}
}

func (c *PubSub) initPing() error {
	//....
	go func() {
		timer := time.NewTimer(time.Minute)
		timer.Stop()

		healthy := true
		timeout := pingTimeout // 1 * time.Second
		for {
			timer.Reset(timeout)
			select {
			case <-c.ping:
				healthy = true
				timeout = pingTimeout
				if !timer.Stop() {
					<-timer.C
				}
			case <-timer.C:
				var healthyErr error

				if healthy {
					healthyErr = c.Ping(ctx)
					healthy = false
				} else {
					healthyErr = errPingTimeout
				}

				if healthyErr != nil {
					c.mu.Lock()
					c.reconnect(ctx, pingErr)
					c.mu.Unlock()

					healthy = true
				} else {
					timeout = c.healthTimeout
				}
			case <-c.exit:
				return
			}
		}
	}()
}
```

**Comment by monkey92t** [comment]
I made changes, any suggestions?
@vmihailenco  Review?

File: pubsub.go [hunk 1]
```diff
@@ -483,26 +495,35 @@ func (c *PubSub) initPing() {
...
 					c.mu.Lock()
-					c.reconnect(ctx, pingErr)
-					healthy = true
+					c.reconnect(ctx, healthyErr)
 					c.mu.Unlock()
+
+					healthy = true
+					timeout = pingTimeout
+				} else {
+					timeout = c.healthTimeout
```
Comment (rbg) [inline pubsub.go:526]: does this make sense -- should we always  use the time that was explicitly set ? 

                                 if healthyErr != nil {
					c.mu.Lock()
					c.reconnect(ctx, healthyErr)
					c.mu.Unlock()
					healthy = true
				} 
				timeout = c.healthTimeout

Comment (monkey92t) [reply pubsub.go:526]: It is the waiting time of the ping command.

**Comment by vmihailenco** [comment]
I've recently refactored similar code for [Bun](https://github.com/uptrace/bun/blob/master/driver/pgdriver/listener.go#L224-L254). As you can see, there is a `Channel` method that accepts an arbitrary number of options. All the code that is required to support it is extracted to the `Channel` struct. It is the 2nd option described in https://github.com/go-redis/redis/issues/1700

Do you think we can do the same thing in go-redis?

File: pubsub.go [hunk 2]
```diff
@@ -271,6 +277,12 @@ func (c *PubSub) Ping(ctx context.Context, payload ...string) error {
 	return err
 }
 
+func (c *PubSub) SetHealthTimeout(t time.Duration) {
+	if t > 0 {
+		c.healthTimeout = t
+	}
+}
+
```
Comment (vmihailenco) [inline pubsub.go:285]: `SetHealthTimeout` only makes sense for `Channel`, but is set on `PubSub`. It is confusing.

**Comment by rbg** [comment]
> I've recently refactored similar code for [Bun](https://github.com/uptrace/bun/blob/master/driver/pgdriver/listener.go#L224-L254). As you can see, there is a `Channel` method that accepts an arbitrary number of options. All the code that is required to support it is extracted to the `channel` struct. It is the 2nd option described in https://github.com/go-redis/redis/issues/1700
> 
> 
> 
> Do you think we can do the same thing in go-redis? We can't just copy-paste the code as is because PubSub in Redis works differently, but we can use the same API.

That is much clearer

**Comment by monkey92t** [comment]
@vmihailenco 

I understand #1700, but there is still a problem with `ChannelSize(size int)`.
If we replace `ChannelSize(size int)` with #1700(`Chan(cfg ChanConfig)`), it will break the compatibility of the v8 version.
#1700 can work in v9.
I want to find a solution for the current v8 version (although not as good as #1700), but it can solve the current problems.

I can remove `SetHealthTimeout` and set it to a fixed value, such as 5 seconds.

**Comment by vmihailenco** [comment]
>I can remove SetHealthTimeout and set it to a fixed value, such as 5 seconds.

I am fine with this and a finding "quick" temp solution in general.

>If we replace ChannelSize(size int) with #1700(Chan(cfg ChanConfig)), it will break the compatibility of the v8 version.
#1700 can work in v9.

I think we should change existing `Channel` to `Channel(opts ...ChannelOption)` (which is backwards compatible). And then change `ChannelSize` to call `Channel`:

```go
func (c *Channel) Channel(opts ...ChannelOption) <-chan *Message {}

func (c *Channel) ChannelSize(size int)  <-chan *Message {
    return c.Channel(redis.WithChannelSize(size))
}
```

All that should be backwards compatible.

**Comment by monkey92t** [comment]
Fixed by #1764

**Pull request closed without merging**